sql/database.go

```diff
@@ -1147,13 +1184,83 @@ func (tx *sqliteTx) Exec(query string, encoder Encoder, decoder Decoder) (int, e
 	return exec(tx.conn, query, encoder, decoder)
 }
 
-type sqliteConn struct {
+// lazyConn is a connection that is acquired lazily from the pool, that is, upon the first
+// query, and released after a certain period of inactivity.
+type lazyConn struct {
 	*queryCache
-	db   *sqliteDatabase
-	conn *sqlite.Conn
+	db      *sqliteDatabase
+	getConn func() *sqlite.Conn
+	eg      errgroup.Group
+	conn    *sqlite.Conn
+	timer   *time.Timer
+	doneCh  chan struct{}
+	connMtx sync.Mutex
+}
+
+func newLazyConn(ctx context.Context, db *sqliteDatabase) *lazyConn {
+	return &lazyConn{
+		queryCache: db.queryCache,
+		db:         db,
+		getConn: func() *sqlite.Conn {
+			return db.getConn(ctx)
+		},
+	}
+}
+
+func (c *lazyConn) ensureConn() *sqlite.Conn {
+	if c.conn != nil {
+		return c.conn
+	}
+
+	c.conn = c.getConn()
+	if c.timer == nil {
+		c.timer = time.NewTimer(c.db.connIdleTimeout)
+		c.eg.Go(func() error {
+			for {
+				select {
+				case <-c.timer.C:
+					// Although TryLock docs say that it's not recommended to use it
+					// in most cases, this use case is justified.
+					// If the mutex is already locked here, this means that an SQL
+					// statement is being executed on the connection, after which
+					// the idle timer will be restarted, or the connection is currently
+					// being released.
+					if c.connMtx.TryLock() {
+						c.releaseConn()
+						c.connMtx.Unlock()
+					}
+				case <-c.doneCh:
+					return nil
+				}
+			}
+		})
+		c.doneCh = make(chan struct{})
```

Comment:
I think it would be clearer if `getCon` inits the timer and the `doneCh` and `releaseCon` stops the timer, sets it to nil and also closes the `doneCh` instead of spreading this across 3 methods.

Reply:
Which `getConn`? `c.getConn` is a function field that's needed so that we don't have to store the context and that's it sole job. `sqliteDatabase.getConn` probably should not be dealing with `doneCh` inside the `lazyConn`.
I don't think that re-creating the timer makes much sense, it may add race with the background goroutine. We could also restart the goroutine, but that's also extra hassle.
And `doneCh` is only used in the background goroutine in `ensureConn` and in `release` where it tells the goroutine to stop when we don't need this particular `lazyConn` anymore.

Reply:
hm... maybe you're right. Would it make sense to invert the if and return early in the case `time != nil`?

Reply:
Fixed